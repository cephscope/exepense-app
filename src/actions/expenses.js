import uuid from 'uuid';
import database from '../firebase/firebase';

// current process:
  // component calls action generator
  // action generator returns object
  // component dispatches object
  // redux store changes

// New process:
  // component calls action generator
  // action genrator generates function
  // component dispatches function (not possible without redux middleware: redux-thunk)
  // function runs (has the ability to dispatch other actions and do whatever it wants e.g. write to firebase)

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

// async function that only works because of Thunk:
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}));
    });
  };
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// async
export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    }).catch((e) => {
      console.log(`something went wrong, here's the:`, e);
    });
  };
};