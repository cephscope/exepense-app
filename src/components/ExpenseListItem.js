import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <Link className="list-item" to={`/edit/${id}`}>
      <div>
        <h3 className="list-item__title">{description}</h3>
        <div className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</div>
      </div>
      <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
    </Link>
);

export default ExpenseListItem;