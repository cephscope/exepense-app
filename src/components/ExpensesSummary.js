import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ visibleExpenses }) => {
  const expenseWord = visibleExpenses.length === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal(visibleExpenses) / 100).format('$0,0.00');

  return (
    <div>
      <h3>{`You're looking at ${visibleExpenses.length} ${expenseWord} totaling ${formattedExpensesTotal}`}</h3>
    </div>
  );
};

const mapStateToProps = (state) => ({
  visibleExpenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);