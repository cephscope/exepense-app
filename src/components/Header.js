import React from 'react';
import { Link, NavLink } from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Fix The Bank</h1>
    <h3>Keep track of your expenses</h3>
    <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard!</NavLink> |
    <NavLink to="/create" activeClassName="is-active">Add Expense Page!</NavLink> |
  </header>
);

export default Header;