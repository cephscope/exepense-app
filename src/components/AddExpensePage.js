import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Expense:</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  };
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);


// DOES THE SAME THING AS ABOVE, BUT IT IS LESS TESTABLE. READ ABOUT mapDispatchToProps // https://github.com/reduxjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options
// const AddExpensePage = (props) => (
//   <div>
//     <h1>Add Expense:</h1>
//     <ExpenseForm 
//       onSubmit={(expense) => {
//         props.dispatch(addExpense(expense));
//         props.history.push('/');
//       }}
//     />
//   </div>
// );

// export default connect()(AddExpensePage);