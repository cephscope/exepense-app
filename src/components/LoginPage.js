import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';


export const LoginPage = ({ startLogin }) => (
  <div>
    <h1>ðŸ¤– Login to me if you want to live ðŸ¤–</h1>
    <h3>(Please log in to start managing your expenses)</h3>
    <button onClick={startLogin}>Login</button>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage)