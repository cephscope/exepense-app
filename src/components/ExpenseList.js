import React from 'react';
import { connect } from 'react-redux'; // imported for all components that need to dispatch action or read from the redux store
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => ( // exporting for test purposes only
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
    {
      props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
         <p>No expenses</p>
        </div>
      ) : (
          props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />;
          })
        )
    }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
}

export default connect(mapStateToProps)(ExpenseList); // connects mapStateToProps to ExpenseList, giving it access to the redux store (in the props)
