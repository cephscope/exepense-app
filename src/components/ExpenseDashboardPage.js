import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
  <div>
    <p></p>
    <ExpensesSummary />
    <p></p>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
export default ExpenseDashboardPage;