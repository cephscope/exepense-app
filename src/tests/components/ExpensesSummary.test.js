import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should return correct message for multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary visibleExpenses={expenses}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should return correct message for one expense', () => {
  const expense = [{
    id: '1',
    description: 'Gum',
    note: '',
    amount: 195,
    createdAt: 0
  }];
  const wrapper = shallow(<ExpensesSummary visibleExpenses={expense}/>);
  expect(wrapper).toMatchSnapshot();
});