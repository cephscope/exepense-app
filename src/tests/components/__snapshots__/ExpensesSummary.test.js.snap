// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correct message for multiple expenses 1`] = `
<div>
  <h3>
    You're looking at 3 expenses totaling $1,141.95
  </h3>
</div>
`;

exports[`should return correct message for one expense 1`] = `
<div>
  <h3>
    You're looking at 1 expense totaling $1.95
  </h3>
</div>
`;
